File: client/src/service/auth.service.ts
import axios from 'axios';

const publicServer = axios.create({
  baseURL: '/',
  headers: {
    'Content-Type': 'application/json',
  },
});

const authorizedServer = axios.create({
  baseURL: '/',
  headers: {
    'Content-Type': 'application/json',
  },
});

publicServer.interceptors.request.use((config) => {
  config['withCredentials'] = true;

  return config;
});

authorizedServer.interceptors.request.use((config) => {
  config['withCredentials'] = true;

  return config;
});

authorizedServer.interceptors.response.use(
  (res) => res,
  async (error) => {
    const originalConfig = error.config;

    if (error.response) {
      if (error.response.status === 401 && !originalConfig._retry) {
        console.log('Refreshing Token');
        originalConfig._retry = true;

        try {
          await fetchNewAccessToken();

          return authorizedServer(originalConfig);
        } catch (_error: any) {
          if (_error.response && _error.response.data) {
            return Promise.reject(_error.response.data);
          }
          return Promise.reject(_error);
        }
      }
    }
  }
);

export const fetchNewAccessToken = async (): Promise<void> => {
  await publicServer.post('api/v1/auth/refresh/');
};

export const getCookie = async (): Promise<void> => {
  await publicServer.post('api/v1/auth/lol/');
};

export const getUser = async (): Promise<void> => {
  const response = await authorizedServer.get(`/api/v1/auth/me/`);

  const responseJson = await response.data;
  console.log(responseJson);
};

export const login = async (): Promise<void> => {
  await publicServer.post(`/api/v1/auth/login/`, {
    username: 'jsmith',
    password: 'test123',
  });
};

File: mobile/lib/service/auth.service.dart
import 'package:dio/dio.dart';
import 'package:mobile/constants/logger.constant.dart';
import 'package:mobile/db/token.db.dart';
import 'package:mobile/exceptions/retry.exception.dart';
import 'package:mobile/models/token/token.model.dart';

class AuthService {
  final TokenDb tokenDb = TokenDb();
  final Dio publicApi = Dio(
    BaseOptions(
      baseUrl: "http://10.0.2.2:5000/",
      connectTimeout: 5000,
      receiveTimeout: 5000,
    ),
  );
  final Dio securedApi = Dio(
    BaseOptions(
      baseUrl: "http://10.0.2.2:5000/",
      connectTimeout: 5000,
      receiveTimeout: 5000,
    ),
  );

  AuthService() {
    securedApi.interceptors.add(
      InterceptorsWrapper(onError: (error, handler) {
        if (error.response?.statusCode == 401) {
          log.i('Access Token Expired');
          _regenerateAccessToken()
              .then((_) => _retry(error.requestOptions))
              .then((response) => handler.resolve(response))
              .catchError(
            (error, stackTrace) {
              log.e(
                "Dio Interceptor Error",
                error,
                stackTrace,
              );
              handler.next(error);
            },
          );
        } else {
          handler.next(error);
        }
      }, onRequest: (options, handler) {
        Token token = tokenDb.fetchActivitiesFromDevice();

        if (token.accessToken.isNotEmpty) {
          options.headers['Authorization'] = "Bearer ${token.accessToken}";
        }

        return handler.next(options);
      }),
    );
  }

  Future<Response<dynamic>> _retry(RequestOptions requestOptions) async {
    final options = Options(
      method: requestOptions.method,
      headers: requestOptions.headers,
    );
    return securedApi.request<dynamic>(
      requestOptions.path,
      data: requestOptions.data,
      queryParameters: requestOptions.queryParameters,
      options: options,
    );
  }

  Future<void> _regenerateAccessToken() async {
    Token token = tokenDb.fetchActivitiesFromDevice();
    Response response = await publicApi.post(
      'api/v1/auth/refresh/',
      data: {
        "refresh_token": token.refreshToken,
      },
    );

    String newAccessToken = response.data['data']['payload']['token'];
    String refreshToken = token.refreshToken;

    tokenDb.saveTokenToDevice(
      Token(
        accessToken: newAccessToken,
        refreshToken: refreshToken,
      ),
    );

    log.i("Refreshed Tokens");
  }

  Future<void> login() async {
    Response response = await securedApi.post(
      'api/v1/auth/login/',
      data: {
        "username": 'jsmith',
        "password": 'test123',
      },
    );

    String accessToken = response.data['data']['payload']['token'];
    String refreshToken = response.data['data']['payload']['refresh_token'];

    tokenDb.saveTokenToDevice(
      Token(
        accessToken: accessToken,
        refreshToken: refreshToken,
      ),
    );
  }

  Future<void> getUser() async {
    try {
      Response response = await securedApi.get(
        'api/v1/auth/me/',
      );
      log.i(response);
    } on RetryException catch (error) {
      log.e(error.message);
    }
  }
}
